browser/page_mapping.go

```diff
@@ -423,70 +423,87 @@ func mapPage(vu moduleVU, p *common.Page) mapping { //nolint:gocognit,cyclop
-				}
-				return nil
-			})
+		// Prepare the environment for the page.on event handler if necessary.
+		if pageOnEvent.prep != nil {
+			if err := pageOnEvent.prep(); err != nil {
+				return fmt.Errorf("initiating page.on('%s'): %w", eventName, err)
+			}
 		}
 
-		onEventPageMetricCalled := func(event common.PageOnEvent) {
-			// The function on the taskqueue runs in its own goroutine
-			// so we need to use a channel to wait for it to complete
-			// since we're waiting for updates from the handler which
-			// will be written to the ExportedMetric.
+		// Queue the event handler in the task queue.
+		// Wait for the handler to complete if necessary.
+		tq := vu.taskQueueRegistry.get(vu.Context(), p.TargetID())
+		queueHandler := func(event common.PageOnEvent) {
 			done := make(chan struct{})
+
 			tq.Queue(func() error {
 				defer close(done)
 
-				mapping := mapMetricEvent(vu, event)
-				_, err := handleEvent(sobek.Undefined(), rt.ToValue(mapping))
+				mapping := pageOnEvent.mapp(vu, event)
+
+				_, err := handleEvent(
+					sobek.Undefined(),
+					rt.ToValue(mapping),
+				)
 				if err != nil {
-					return fmt.Errorf("executing page.on('metric') handler: %w", err)
+					return fmt.Errorf("executing page.on('%s') handler: %w", eventName, err)
 				}
 
 				return nil
 			})
-			<-done
-		}
 
-		var mapHandler func(common.PageOnEvent)
-		switch eventName {
-		case common.EventPageConsoleAPICalled:
-			mapHandler = onEventPageConsoleAPICalled
-		case common.EventPageMetricCalled:
-			mapHandler = onEventPageMetricCalled
-		default:
-			return fmt.Errorf("unknown page event: %q", eventName)
+			if pageOnEvent.wait {
+				<-done
+			}
 		}
 
-		if eventName == common.EventPageMetricCalled {
-			// Register a custom regex function for the metric event
-			// that will be used to check URLs against the patterns.
-			// This is needed because we want to use the JavaScript regex
-			// to comply with what users expect when using the `tag` method.
-			_, err := rt.RunString(`
-				function _k6BrowserCheckRegEx(pattern, url) {
-					let r = pattern;
-					if (typeof pattern === 'string') {
-						r = new RegExp(pattern);
-					}
-					return r.test(url);
+		// Run the the event handler in the task queue to ensure that
+		// the handler is executed in the event loop.
+		return p.On(eventName, queueHandler) //nolint:wrapcheck
+	}
+}
+
+// prepK6BrowserRegExChecker is a helper function to check the regex pattern
+// on Sobek runtime. Unlike Go's regexp package, Sobek's runtime checks
+// regex patterns using JavaScript's regular expression features.
+func prepK6BrowserRegExChecker(rt *sobek.Runtime) func() error {
+	return func() error {
+		_, err := rt.RunString(`
+			function _k6BrowserCheckRegEx(pattern, url) {
+				let r = pattern;
+				if (typeof pattern === 'string') {
+					r = new RegExp(pattern);
 				}
-			`)
-			if err != nil {
-				return fmt.Errorf("evaluating regex function: %w", err)
+				return r.test(url);
```

Comment:
I just remembered that this isn't correct. Happy to change this myself in another PR, but we only accept JS regex types not strings. So this should just be:

```suggestion
				return pattern.test(url);
```

I think in the future we will need to accept glob patterns and maybe even exact string matches.

I don't think we will be able to accept a string pattern that is used in `new RegExp` due to working with Sobek.

If a user uses `'/^https:\/\/example\.com\/checkout\/[0-9a-z]+$/'` Sobek removes the delimieters `/^https://example.com/checkout/[0-9a-z]+$/`, which makes it impossible to work with in `new RegExp`.